File: python-algorithm/ds_5_tree/tree_1_4_path_recurrent/L337-m.py
#--coding:utf-8--


class Solution:

    dic = {} # 记忆化搜索

    def rob(self, root: TreeNode) -> int:
        if root == None:
            return 0
        if root in self.dic:
            return self.dic[root]
        val1 = root.val
        if root.left:
            val1 += self.rob(root.left.left) + self.rob(root.left.right)
        if root.right:
            val1 += self.rob(root.right.left) + self.rob(root.right.right)
        val2 = self.rob(root.left) + self.rob(root.right)
        res = max(val1, val2)
        self.dic[root] = res
        return res


# 动态规划  没咋看懂
# ls表示偷左子树能带来的最大收益，ln表示不偷左子树能带来的最大收益，rs、rn同理
class Solution:
    def rob(self, root: TreeNode) -> int:
        def _rob(root):
            if not root:
                return 0, 0
            ls, ln = _rob(root.left)
            rs, rn = _rob(root.right)
            return root.val + ln + rn, max(ls, ln) + max(rs, rn)
        return max(_rob(root))

File: python-algorithm/ds_5_tree/tree_1_4_path_recurrent/L337-m.cpp


//间隔遍历二叉树

#include <iostream>
#include <vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

//暴力递归 最后个案例没过
class Solution {
public:
    int rob(TreeNode* root) {
        if (root == NULL) return 0;
        int val1 = root->val;
        if (root->left != NULL) val1 += rob(root->left->left) + rob(root->left->right);
        if (root->right != NULL) val1 += rob(root->right->left) + rob(root->right->right);
        int val2 = rob(root->left) + rob(root->right);
        return max(val1, val2);
    }
};

//记忆化搜索，时间效率提高了，空间复杂度上去了
#include <unordered_map>
class Solution1 {
public:
    unordered_map<TreeNode*, int > um;
    int rob(TreeNode* root) {
        if (root == NULL) return 0;
        if(um.find(root) != um.end())
            return um[root];
        int val1 = root->val;
        if (root->left != NULL) val1 += rob(root->left->left) + rob(root->left->right);
        if (root->right != NULL) val1 += rob(root->right->left) + rob(root->right->right);
        int val2 = rob(root->left) + rob(root->right);
        int res = max(val1, val2);
        um[root] = res;
        return res;
    }
};

//类似动态规划，没看懂
class Solution2 {
public:
    int rob(TreeNode* root) {
        vector<int> result = robInternal(root);
        return max(result[0], result[1]);
    }

    vector<int> robInternal(TreeNode* root) {
        if (root == NULL) return vector<int>(2,0);
        vector<int> result(2,0);

        vector<int> left = robInternal(root->left);
        vector<int> right = robInternal(root->right);

        result[0] = max(left[0], left[1]) + max(right[0], right[1]);
        result[1] = left[0] + right[0] + root->val;
        return result;
    }


};